### PR TITLE
[xpu] Enable GPU ARCH filter to FlashAttention Unit Test in CI

### PR DESCRIPTION
**Pull request by LuFinch**
Currently, Pytorch XPU is using **Intel Data Center GPU Max** in CI/CD. 

However, SYCLTLA(SYCL Templates for Linear Algebra) only formally supports B580/B60 now. The LTS2 driver on **Intel Data Center GPU Max** may cause accuracy issue in SYCLTLA.

Hence, XPU will skip Unit Test for the kernels imlemented by SYCLTLA on **Intel Data Center GPU Max** and runs the UT on B60 instead. See https://github.com/pytorch/pytorch/pull/174188

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by etaf — APPROVED** [review]

File: torch/_inductor/codegen/xpu/xpu_env.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+@clear_on_fresh_cache
+@functools.lru_cache(1)
+def get_xpu_arch() -> str | None:
+    arch_code2name = {
+        13136561920: "Xe12",
+        21479031808: "Xe20",
+        21483225088: "Xe20",
+    }
+    try:
+        arch_code = torch.xpu.get_device_capability()["architecture"]
```
Comment (EikanWang) [inline torch/_inductor/codegen/xpu/xpu_env.py]: @guangyey , do you know the timeline that SYCL provides the arch info?

File: test/test_transformers.py [hunk 2]
```diff
@@ -4576,7 +4577,7 @@ class TestSDPAXpuOnly(NNTestCase):
     Mostly migrate from TestSDPACudaOnly in test/test_transformers.py
     """
 
-    PLATFORM_SUPPORTS_XPU_FLASH_ATTENTION = torch.xpu.is_available() and torch._C._is_flash_attention_available()
+    PLATFORM_SUPPORTS_XPU_FLASH_ATTENTION = torch.xpu.is_available() and torch._C._is_flash_attention_available() and PLATFORM_SUPPORTS_SYCLTLA
```
Comment (EikanWang) [inline test/test_transformers.py]: @LuFinch , in terms of `PLATFORM_SUPPORTS_SYCLTLA`, we can replace it with `XE2_LATER`.

**Review by EikanWang — CHANGES_REQUESTED** [review]
@LuFinch, @etaf, I'm somewhat conservative about introducing `PLATFORM_SUPPORTS_SYCLTLA`, as we intend to use the variable for a long time. So, I would like to have a serious discussion to decide if the variable is a real need.

File: torch/_inductor/codegen/xpu/xpu_env.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+@clear_on_fresh_cache
+@functools.lru_cache(1)
+def get_xpu_version() -> str | None:
+    # string of version, like 20250101
+    try:
+        xpu_version = torch.version.xpu
+        return xpu_version
+    except Exception:
+        log.exception("Error getting xpu version")
+        return None
```
Comment (guangyey) [inline torch/_inductor/codegen/xpu/xpu_env.py]: Where is this function used? I recommend using `torch.version.xpu` directly.

Comment (guangyey) [inline torch/_inductor/codegen/xpu/xpu_env.py]:
```suggestion
    if not torch.version.xpu:
        log.exception("Error getting xpu version")
    return torch.version.xpu
```

Comment (etaf) [reply torch/_inductor/codegen/xpu/xpu_env.py]: This PR is a subset of #161940, we would like to land this part first to make the transformer test runs on Client GPU CI machine.

File: torch/_inductor/codegen/xpu/xpu_env.py (hunk 1 above)
Comment (guangyey) [reply torch/_inductor/codegen/xpu/xpu_env.py]: The earliest possible timeline would be version 2026.2.

File: torch/testing/_internal/common_xpu.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+from torch._inductor.codegen.xpu.xpu_env import get_xpu_arch
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]: This doesn't make sense that this file depends on torch._inductor.

File: torch/_inductor/codegen/xpu/xpu_env.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import torch
+from torch._inductor.utils import clear_on_fresh_cache
+
+
+log = logging.getLogger(__name__)
+
+
+@clear_on_fresh_cache
+@functools.lru_cache(1)
+def get_xpu_arch() -> str | None:
```
Comment (guangyey) [inline torch/_inductor/codegen/xpu/xpu_env.py]: Let's not define this arch by ourselves. I recommend us to detect like rocm using `device_id`, which is a stable symbol.

File: torch/testing/_internal/common_xpu.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+
+
+PLATFORM_SUPPORTS_SYCLTLA: bool = get_xpu_arch() == "Xe20"
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]: We should not add a xpu-specific file here. Refer to https://github.com/pytorch/pytorch/pull/177636, we should add it to `common_gpu.py`

Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]: 1. Use `torch.xpu.get_device_properties().device_id` to detect Xe2 arch as `device_id` is a stable number defined in spec.
2. Here we must use `LazyVal`, otherwise it would trigger XPU runtime, which is not the situation we expect.

Comment (guangyey) [reply torch/testing/_internal/common_xpu.py]: It's fine to add `common_xpu.py` here. You’ve convinced me

File: test/test_transformers.py (hunk 2 above)
Comment (LuFinch) [reply test/test_transformers.py]: Sure, Done.

File: torch/_inductor/codegen/xpu/xpu_env.py (hunk 3 above)
Comment (LuFinch) [reply torch/_inductor/codegen/xpu/xpu_env.py]: I remove this function due to it is unrelated to this PR. Please enable it in your cutlass PR alone. @etaf

File: torch/_inductor/codegen/xpu/xpu_env.py (hunk 5 above)
Comment (LuFinch) [reply torch/_inductor/codegen/xpu/xpu_env.py]: Done

File: torch/testing/_internal/common_xpu.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    if TEST_XPU:
+        return not IS_WINDOWS and Xe2_Or_Later
+    return False
+
+PLATFORM_SUPPORTS_FLASH_ATTENTION_XPU: bool = LazyVal(lambda: evaluate_platform_supports_flash_attention())
+
+# Importing this module should NOT eagerly initialize XPU
+if not XPU_ALREADY_INITIALIZED_ON_IMPORT:
+    if torch.xpu.is_initialized():
+        raise AssertionError("XPU should not be initialized on import")
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]:
```suggestion
        raise AssertionError("XPU should not be initialized on import")

```
Need a blank line.

File: torch/testing/_internal/common_xpu.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+_CODENAME_TO_ARCH = {
+    XPUCodename.PVC: XPUArch.Xe,
+    XPUCodename.BMG: XPUArch.Xe2,
+}
+
+
+@functools.lru_cache(1)
+def get_xpu_device_id() -> int | None:
+    try:
+        return torch.xpu.get_device_capability()["device_id"]
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]: In my understanding, this API would never raise an exception. `device_id` is always meaningful and returns `0` if undefined.

File: torch/testing/_internal/common_xpu.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+
+class XPUCodename(enum.Enum):
+    PVC = "PVC"  # Intel® Data Center GPU Max Series
+    BMG = "BMG"  # Intel® Arc™ Pro Battlemage Graphics
+
+
+class XPUArch(enum.Enum):
+    Xe = "Xe"  # Xe HPC
+    Xe2 = "Xe2"
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]:
```suggestion
    Unknow = "Unknow"
    Xe = "Xe"  # Xe HPC
    Xe2 = "Xe2"
```
In this PR, `Xe` is unused. If you prefer to define it, I suggest adding `Unknown` to make sure the XPUArch supports the comparison operation easily.

File: torch/testing/_internal/common_xpu.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+@functools.lru_cache(1)
+def get_xpu_arch() -> XPUArch | None:
+    codename = get_xpu_codename()
+    if codename is None:
+        return None
+    return _CODENAME_TO_ARCH.get(codename)
+
+
+Xe2_Or_Later = LazyVal(
+    lambda: torch.xpu.is_available() and get_xpu_arch() in (XPUArch.Xe2,)
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]:
```suggestion
    lambda: torch.xpu.is_available() and get_xpu_arch() >= XPUArch.Xe2
```

File: torch/testing/_internal/common_xpu.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        log.exception("Error in getting xpu device_id.")
+        return None
+
+
+@functools.lru_cache(1)
+def get_xpu_codename() -> XPUCodename | None:
+    device_id = get_xpu_device_id()
+    if device_id is None:
+        return None
+    return _DEVICE_ID_TO_CODENAME.get(device_id)
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]:
```suggestion
    device_id = torch.xpu.get_device_capability()["device_id"]
    return _DEVICE_ID_TO_CODENAME.get(device_id, None)
```

File: torch/testing/_internal/common_xpu.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+
+@functools.lru_cache(1)
+def get_xpu_arch() -> XPUArch | None:
+    codename = get_xpu_codename()
+    if codename is None:
+        return None
+    return _CODENAME_TO_ARCH.get(codename)
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py]:
```suggestion
    codename = get_xpu_codename()
    return _CODENAME_TO_ARCH.get(codename, XPUArch.Unknow)
```

File: torch/testing/_internal/common_xpu.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+
+Xe2_Or_Later = LazyVal(
```
Comment (guangyey) [inline torch/testing/_internal/common_xpu.py:70]:
```suggestion
Xe2OrLater = LazyVal(
```
Is this style better? This is not a must-do.

File: torch/testing/_internal/common_xpu.py (hunk 9 above)
Comment (LuFinch) [reply torch/testing/_internal/common_xpu.py]: Done

File: torch/testing/_internal/common_xpu.py (hunk 10 above)
Comment (LuFinch) [reply torch/testing/_internal/common_xpu.py]: Done.

File: torch/testing/_internal/common_xpu.py (hunk 9 above)
Comment (LuFinch) [reply torch/testing/_internal/common_xpu.py]: To make the eumn class can be compared directly, I make the value as int instead of string.

File: torch/testing/_internal/common_xpu.py (hunk 11 above)
Comment (LuFinch) [reply torch/testing/_internal/common_xpu.py]: done

File: torch/testing/_internal/common_xpu.py (hunk 12 above)
Comment (LuFinch) [reply torch/testing/_internal/common_xpu.py]: Done

File: torch/testing/_internal/common_xpu.py (hunk 13 above)
Comment (LuFinch) [reply torch/testing/_internal/common_xpu.py:70]: The `2` and `O` is kind of confusing. I prefer keep it as it is.

**Review by guangyey — APPROVED** [review]
Thanks for the update!

File: test/test_transformers.py [hunk 14]
```diff
@@ -5155,6 +5156,7 @@ def test_causal_variants(self, device, causal_variant: CausalVariant, shape: lis
         [(16, 16, 128, 128, 16), (16, 16, 128, 256, 32), (16, 16, 256, 128, 32), (1, 1, 23, 56, 15)],
     )
     @skipIfTorchDynamo("This function already calls torch.compile.")
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION_XPU, "XPU Flash Attention is not supported")
```
Comment (guangyey) [inline test/test_transformers.py]: May I know why the test case doesn't have the decorator `PLATFORM_SUPPORTS_XPU_FLASH_ATTENTION` previously?

Comment (LuFinch) [reply test/test_transformers.py]: The Pytorch main branch is using SYCLTLA-v0.6 branch and SYCLTLA legacy API for FlashAttention now and there is no issue for PVC.

We want to move to SYCLTLA-v0.7 branch and use new API, but they depends on new version IGC. However, the PVC LTS2 driver's IGC is too old so that we need to skip PVC test cases and use BMG CI to run these test cases instead.

Comment (guangyey) [reply test/test_transformers.py]: Thanks.

File: test/test_transformers.py [hunk 15]
```diff
@@ -4883,7 +4883,7 @@ def test_onednn_attention_mask_vs_math(
 
         self.assertEqual(actual.float(), math_ref, atol=tol.atol, rtol=tol.rtol)
 
-    @unittest.skipIf(not PLATFORM_SUPPORTS_XPU_FLASH_ATTENTION, "XPU Flash Attention is not supported")
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION_XPU, "XPU Flash Attention is not supported")
```
Comment (etaf) [inline test/test_transformers.py]: Let's use `skipXPUIf`.

File: torch/testing/_internal/common_xpu.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+class XPUCodename(enum.Enum):
+    PVC = "PVC"  # Intel® Data Center GPU Max Series
+    BMG = "BMG"  # Intel® Arc™ Pro Battlemage Graphics
+
+
+class XPUArch(enum.Enum):
+    Unknown = 0
+    Xe = 1  # Xe HPC
+    Xe2 = 2
+
```
[the remark this answers is not in the document]
Comment (etaf) [reply torch/testing/_internal/common_xpu.py]: Please fix the issue.

**Review by etaf — CHANGES_REQUESTED** [review]

File: torch/testing/_internal/common_xpu.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (LuFinch) [reply torch/testing/_internal/common_xpu.py]: Done

File: test/test_transformers.py (hunk 15 above)
Comment (LuFinch) [reply test/test_transformers.py]: `skipXPUIf` can't accept condition to check `XeArch`. I change to `@unittest.skipIf(TEST_XPU and not PLATFORM_SUPPORTS_FLASH_ATTENTION_XPU` so that CUDA/ROCM won't be skipped.

Comment (etaf) [reply test/test_transformers.py]: we can use `skipXPUIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION_XPU, "XPU Flash Attention is not supported")` like https://github.com/pytorch/pytorch/pull/161940/changes#diff-6c0fb46de365a3c6711e9a786a88226556aaa5035346772ff2afcb1e5bfe8262R290

Comment (LuFinch) [reply test/test_transformers.py]: thanks for your remind.  Done.

**Review by etaf — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]
LGTM
